### PR TITLE
signpipe stumbles over interrupted rrsets

### LINKED ISSUE #5227: reordering RRs in bind zone file makes pdns return invalid RRSIGs on AXFR
With pdns 4.0.3 on Debian Jessie from powerdns’ repositories, the following part of the bind zone file makes pdns return one (valid) rrsig on direct query, but two (invalid) rrsigs on axfr:

    imap 		    60   IN A 	    138.201.157.89
    imap 		    60   IN AAAA    2a01:4f8:c17:4920::2
    imap		    60   IN A	    195.191.196.74
    imap		    60   IN AAAA    2a02:2918:1017:74::1

On direct query, one rrsig is returned per rrtype:

    % dig +dnssec imap.zombofant.net @ns0.q-ix.net
     
    ;; ANSWER SECTION:
    imap.zombofant.net.	60	IN	A	195.191.196.74
    imap.zombofant.net.	60	IN	RRSIG	A 10 3 60 20170420000000 20170330000000 13103 zombofant.net. n1v74KKI+lsjDmB4z+nmmzMrvMC4vp3qKKumjqGzMcElCdR6uioU2Sb/ o4STgxgX/x9JF+SuWnOUW2No5amWJAcR6rO7dCkP9GOl8wAWZZfCsqSp X+NttFkbwDaj0woUOICjOa8oCAg4gDG0wPteltMrkMGJq9gqYM56hW54 V1o=
    imap.zombofant.net.	60	IN	A	138.201.157.89
    
    % dig +dnssec imap.zombofant.net aaaa @ns0.q-ix.net
    […]
    ANSWER SECTION:
    imap.zombofant.net.	60	IN	RRSIG	AAAA 10 3 60 20170420000000 20170330000000 13103 zombofant.net. g1P06dDgMoOAC/POgUz8vd0VppZe5x40iElZ+WgtFDPFD/00dKCiw8J1 tVM3Wr4qbvr8LpGP/4qX6SYb7UhOFLpspchpKNo0Hq8/RKdmv5zSfCOL +i/Ih9mVN1JleLEutb9rOqdH4GjJjxIVBLhI5jcv/gLg55xuWbqjvtTH BiY=
    imap.zombofant.net.	60	IN	AAAA	2a02:2918:1017:74::1
    imap.zombofant.net.	60	IN	AAAA	2a01:4f8:c17:4920::2

On AXFR, two rrsigs different from the other are returned:

    % dig axfr zombofant.net @ns0.q-ix.net
    […] 
    imap.zombofant.net.	60	IN	A	138.201.157.89
    imap.zombofant.net.	60	IN	RRSIG	A 10 3 60 20170420000000 20170330000000 13103 zombofant.net. aH3IBML6QALph+bmYcwKeBXdbMcQzD+ktT47iyX0NtoOANIwKCTBhblA LCSjXOfdBQ3uKsmTE2F+4NlEW4uwHCD0Noc+JtQThNxqJBi2tARkpNze l3XjZzRuP1DxKKx0EiRWUxT/WJhn2AL2Jp4aa6rH4+F9A/XTb/oJ4Oti sn8=
    imap.zombofant.net.	60	IN	AAAA	2a02:2918:1017:74::1
    imap.zombofant.net.	60	IN	RRSIG	AAAA 10 3 60 20170420000000 20170330000000 13103 zombofant.net. tfnJp4haWbfg7WBRtrkSI6qRtQ91bv92TI/jcDyhPjJENHwkSvBjawr3 S2KBoBb0/6OAKSo23rJO6dQVEL5JUZw/7nr5adtzHJ8tMdFxlkILpvk9 Eb8ZibR1ntLrtWMmst9ZgkMYnUCnn2KUi3jJqildc78ysmOk9y9ve4q0 ShE=
    […]
    imap.zombofant.net.	60	IN	AAAA	2a01:4f8:c17:4920::2
    imap.zombofant.net.	60	IN	RRSIG	AAAA 10 3 60 20170420000000 20170330000000 13103 zombofant.net. VEYGilbqBZvXCjk2duFgHMJ89Y0DA2CO3q6QZnvXsYtF8fjJxgOihzML BUxq9uuFPzFUstsBMdg51V7c67asPDsi98ZwVcWRHr6JI4Wkjm+Ld5Ll +LxC+KK36LaKC60q7b+PIIwFEM4Q5odiB4UJhMGS6HSP3gdAUrMD4huo GVM=
    […]
    imap.zombofant.net.	60	IN	A	195.191.196.74
    imap.zombofant.net.	60	IN	RRSIG	A 10 3 60 20170420000000 20170330000000 13103 zombofant.net. TMAWaUqQ0Ki9haQ7tHgqyg9yPmenJDNF3EAFg1wuyiSVg6ZBUgVmBGMz 0VCyBUfn6Wb7fS+bfyHQTR9g/Nlwfl4wxjAGfXSho649C0yWfRab7MPj thM6dTQDj2U77iuGXDSv1WI/FC4Sxd6H5tCtPl9/akd65XQyXTEVMxnN pE8=

Sorting the records in the bind zone by record type fixes the issue:

    imap 		    60   IN A 	    138.201.157.89
    imap		    60   IN A	    195.191.196.74
    imap 		    60   IN AAAA    2a01:4f8:c17:4920::2
    imap		    60   IN AAAA    2a02:2918:1017:74::1

### PR DESCRIPTION
**Pull request by mind04**
### Short description
signpipe get confused if it receives a rrset in multiple chunks. This is resulting in extra and broken RRSIG records in axfr output.
fixes #5227 

### Checklist
be checked for the Pull Request to be accepted -->
I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled and tested this code
- [ ] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

### CODE REVIEW

File: pdns/tcpreceiver.cc [hunk 1]
```diff
@@ -809,6 +809,12 @@ int TCPNameserver::doAXFR(const DNSName &target, shared_ptr<DNSPacket> q, int ou
     }
   }
 
+  // Group records by name and type, signpipe stumbles over interrupted rrsets
+  sort(zrrs.begin(), zrrs.end(), [](const DNSZoneRecord& a, const DNSZoneRecord& b) {
+    if (a.dr.d_name != b.dr.d_name) return a.dr.d_name < b.dr.d_name;
```
Comment (ahupowerdns) [inline pdns/tcpreceiver.cc]: I've been burned by messing this up, I suggest: return std::tie(a.dr.d_name, b.dr.d_name) < std::tie(a.dr.d_type, b.dr.d_type); I know you got the code right, but when I have to debug something I have to run the cleverness on my brain and the idiomatic std::tie construct I don't have to think about

File: regression-tests.nobackend/tinydns-data-check/expected_result [hunk 2]
```diff
@@ -1,5 +1,5 @@
 17a1219ca79020c1c67ad173a291d3c8  ../regression-tests/zones/example.com
-b4feb02bd50f7b78cd5c40c7a560019c  ../regression-tests/zones/test.com
+fe49d2784b1bcc3b91ddd5619f0b6cc1  ../regression-tests/zones/test.com
```
Comment (Habbie) [inline regression-tests.nobackend/tinydns-data-check/expected_result]: If you change this, please also regenerate the tinydns data files. I'll do it for you now.

**Comment by Habbie** [comment]
After #5229 is merged, please rebase this.

**Comment by Habbie** [comment]
@mind04 can you please rebase? #5229 was merged

**Comment by mind04** [comment]
Done

**Pull request merged**